Worker thread pool and command line changes

ghjm opened the pull request:
- Change plugin worker interface to use a thread pool and thread-safe queue
  	- This isolates plugins from the asyncio event loop, and makes them simpler.
	- This breaks all existing plugins.
- Change recv() to return the whole message, not just the raw_payload
	- This also breaks all existing plugins.
- Add an EOF message type that indicates all response data has been sent
- Create a category of 'ephemeral' commands that will exit when EOF is received
- Use /tmp/receptor instead of /var/lib/receptor for ephemeral commands
- Connect to localhost:8888 by default for ping, status and send
- Change default ping behavior to count of 4 with 1-second intervals

matburt commented on receptor/controller.py:
```diff
@@ -37,8 +37,7 @@ def add_peer(self, peer):
     async def recv(self):
-        inner = await self.receptor.response_queue.get()
-        return inner.raw_payload
+        return await self.receptor.response_queue.get()
```
In the future this is probbly going to be a variant of `Message` but this is fine.

ghjm replied:
Matthew 6:34

matburt commented on receptor/entrypoints.py:
```diff
@@ -106,7 +106,13 @@ def run_as_send(config):
+            if message.message_type == 'response':
+                print("{}".format(message.raw_payload))
+            elif message.message_type == 'eof':
+                break
```
Can we add a `.debug(...)` here for the EOF message?

matburt replied:
Maybe even a higher verbosity, so the user knows that it actually finished rather than having given up for some reason.

ghjm replied:
Matthew 6:34

matburt commented on receptor/config.py:
```diff
@@ -403,6 +417,12 @@ def parse_options(self, args):
+        # If we did not get a data_dir from anywhere else, use a default
+        if self._config_options['default_data_dir'].value is None:
+            if hasattr(self._parsed_args, 'is_ephemeral') and self._parsed_args.is_ephemeral:
+                self._config_options['default_data_dir'].value = '/tmp/receptor'
```
are we relying on tmpfs cleanup here rather than attempting to remove the connection directories after the ephemeral commands go away?

ghjm replied:
I planned to write tmp directory cleanup for ephemeral commands, but first wanted to get some feedback on whether people even think this ephemeral command thing is a good idea.  I'll go ahead and write the tmp dir cleanup.

matburt replied:
I'll short circuit that. It's *definitely* a good thing.

ghjm replied:
Done

matburt commented on receptor/config.py:
```diff
@@ -91,6 +96,14 @@ def __init__(self, args=None):
             value_type='bool',
             hint='Emit debugging output.',
         )
+        default_max_workers = min(32, os.cpu_count() + 4)
```
Can we add this information (plus how many workers are consumed) to the capabilities broadcast?

ghjm replied:
WIlco

ghjm replied:
Added the number of configured threads.  It's not a simple matter to report how many threads are active, because ThreadPoolExecutor doesn't have a published API to get this information, and also re-uses idle threads (so a thread that is "active" isn't necessarily actually doing work).

matburt commented on receptor/work.py:
```diff
@@ -55,18 +61,37 @@ def remove_work(self, env):
-                )
-                await self.receptor.router.send(enveloped_response)
+            response_queue = queue.Queue()
+            work_exec = self.thread_pool.submit(action_method, inner_env, self.receptor.config.plugins.get(namespace, {}), response_queue)
```
why this instead of `loop.run_in_executor` ?

matburt replied:
I reckon the answer to that is because the thread is not `await`able

matburt replied:
Then we could use a `future` or `gather` below and might not need the loop that follows this.

matburt replied:
check how we do this on the buffers: https://github.com/project-receptor/receptor/blob/devel/receptor/buffers/file.py#L54

ghjm replied:
It looks like `loop.run_in_executor` is meant for the case where we have a queue, and we want aio to consume messages from the queue and execute each one in a thread.  Our situation is that we have a queue that is _producing_ messages _from_ a thread, and we want a bridge pattern where those messages get placed on an aio queue.  I don't see any aio built-ins that are intended or useful for this purpose.  Of course I'm willing to be educated...

matburt commented:
We need to update the merge target for this PR to `devel` from `master`

ghjm commented:
Added more stuff to this PR:

- Updated file buffers so manifests are kept in the node directory; otherwise, tmpdir cleanup doesn't work.  (And this seems to be the intent - I think the current behavior is just a mistake.)
- I was using an incorrect method to mark receptor instances as ephemeral; I have fixed this.
- Ephemeral instances now get removed from routing, rather than going to cost=100, at shutdown.
- The output of `receptor status`, while still human readable, is also now valid YAML so it can be machine parsed if desired.
- Security fix to avoid receptor nodes being able to call arbitrary methods in worker modules.  You now have to explicitly mark a method as receptor exportable.

adamruzicka commented:
When I make a silly mistake in the plugin code (for example type `asnycio` instead of `asyncio`) the execution silently dies when hitting it and the error does not get propagated to the console.

Logs: https://p.teknik.io/qgHQD

ghjm commented:
The exception issue is fixed @adamruzicka.
